packages/roosterjs-editor-dom/lib/edit/adjustInsertPosition.ts

```diff
@@ -160,28 +159,11 @@ function adjustInsertPositionForStructuredNode(
                     : PositionType.After
             );
         }
-    } else if (tag == 'TABLE' && trNode) {
-        // When inserting a table into a table, if these tables have the same column count, and
-        // current position is at beginning of a row, then merge these two tables
-        let newTable = new VTable(<HTMLTableElement>rootNodeToInsert);
-        let currentTable = new VTable(<HTMLTableCellElement>tdNode);
+    }
 
-        if (
-            currentTable.col == 0 &&
-            tdNode == currentTable.getCell(currentTable.row || 0, 0).td &&
-            newTable.cells?.[0] &&
-            newTable.cells[0].length == currentTable.cells?.[0].length &&
-            isPositionAtBeginningOf(position, tdNode)
-        ) {
-            if (
-                getTagOfNode(rootNodeToInsert!.firstChild) == 'TBODY' &&
-                !rootNodeToInsert!.firstChild?.nextSibling
-            ) {
-                unwrap(rootNodeToInsert!.firstChild!);
-            }
-            unwrap(rootNodeToInsert!);
-            position = new Position(trNode, PositionType.After);
-        }
+    if (tag == 'TABLE' && trNode) {
+        pasteTable(root, nodeToInsert, position, range);
+        return null;
```

Comment:
The behaviour is insertNode is kept consistent. The two options I see is add a check for non null position before inserting or add a check for table inside table insertion before inserting. I do not see how the second one is more efficient.

Reply:
As a general function, insertNode() should know as few details of the adjusting result as possible. So if we always return a valid position, it doesn't need to check whether it is null and always do the same handling. Then we keep all those details in a smaller scope, such as in what case insertNode() need to insert.

Reply:
How do I keep them on a smaller scope without affecting the whole behaviour? I still do not understand why it is more efficient.

Reply:
Let's see two examples:

example 1:
```ts
function foo(): Position | null {
  if (...) {
     return {...};
  } else {
     return null;
  }
}

const result = foo();
if (result) {
  // do something with the result
} else {
  // result is null, do something else
}
```

Example 2:
```ts
function foo(): Position {
  if (...) {
     return {...};
  } else {
     return { ... } ; // Still return a valid position, so when call the function, no need to know the detail that when return object, and when return null
  }
}

const result = foo();
// Here we know result must not be null, so we can always do the samething to result. function foo() hides the implementation details so caller doesn't need to care it.
```

Reply:
Sure I understand that part, and I thank you for the explanation, but what I do not understand is why the alternative is better or even how you are suggesting I implement the alternative.

If I do return a valid position there will be an unwanted insertion, I need to stop that insert or modify what is inserted to be nothing.

Reply:
When the node to insert is an empty DocumentFragment, the insert operation will be an no-op, so nothing special need to be done there.

About "why the alternative is better", it is because if a function can always return a valid value and can be processed used by one single way, we need to do so. Otherwise each place this function is called, we always need to handle the result in "if ... else". Sometimes that necessary, but in this case, since we can handle it in one single way, why still give a "null" result and let the caller handle it?

Reply:
If you are familiar with C++ (or any other OO language), you may be familiar with virtual function. A virtual function can have different implementation in each sub class. As a caller, we don't need to care what implementation it is, but just always call it. That means, with the same interface, we can always do the same handling. Here it is the same.

Reply:
This tread started with the suggestion to return a valid position and return an empty fragment to insert. I understand why you made that suggestion and agree it would be better, but I still do not know how it is possible to make the fragment empty, with the current way insertNode works, in a manner than does not use position. As position is the only thing that adjustInsertPosition returns and there is where the pasting is handled.

Reply:
I used moveChildNodes again to remove the children of the node to insert. I honestly did not know that nodeToInsert was a reference and not a copy to the node to insert.

Reply:
To make fragment empty, you just need to call moveChildNodes() to clear it. And actually you need to do that even you return null in order to keep consistent with other cases. After insert, we always expect the fragment to be empty